### DISCUSSION TITLE
Is it possible to SSH into code-server on VS Code

### DISCUSSION THREAD

**Discussion by AyaanZaveri** [coder/code-server#3961] [Q&A]
I'm running code-server on Railway, but I would really love to have it synced. So, is there a way to SSH into a code-server.

**Comment by jsjoeio** [comment]
I don't think so sadly :( We would have to implement this ourselves. See https://github.com/cdr/code-server/issues/1315

**Comment by lg-nineoneninethreenine** [comment]
> I'm running code-server on Railway, but I would really love to have it synced. So, is there a way to SSH into a code-server.

Same thing which I want but I realized, Railway is running in a container, not a VM, so might be possible.

**Reply by jsjoeio** [reply]
Ooo good point. If you figure out a solution, let us know!

**Reply by lg-nineoneninethreenine** [reply]
Not any kind of solution yet but I will try checking about that. I actually tried SSHing through VSCode, but since these run on servers, I will try finding something.

**Reply by AyaanZaveri** [reply]
Thanks!

**Reply by AyaanZaveri** [reply]
There is something called [chisel](https://github.com/jpillora/chisel) that GitPod has used before [here](https://github.com/corneliusludmann/gitpod-playground/tree/sshd-over-chisel).

**Reply by lg-nineoneninethreenine** [reply]
> There is something called [chisel](https://github.com/jpillora/chisel) that GitPod has used before [here](https://github.com/corneliusludmann/gitpod-playground/tree/sshd-over-chisel).

Sorry for the late reply, but it is only for Gitpod. Modifying the dockerfile seems(and IS) futile to the deployment as when I did it, it always failed. @jsjoeio do you have any kind of solution to solve the dockerfile deployment?

**Reply by lg-nineoneninethreenine** [reply]
Moreover, I am a user of Gitpod and tried SSHing through that too. But, code-server(deployed) and Gitpod runs on SERVERS, not home computers. So, the IP addresses change whenever you access it, but I think this doesn't happen on code-server. Moreover, code-server and Gitpod has some STRIKING SIMILARITIES. So, `chisel` can work on Gitpod, but it can't work on code-server because it needs a perfect image for that. @jsjoeio can you try making a simple image with `chisel` inside that?

**Reply by lg-nineoneninethreenine** [reply]
Chisel doesn't work for me. Try for yourself in a burner repo. I use Windows so it didn't happen.

**Reply by AyaanZaveri** [reply]
I tried using ngrok but the password is never right! I tried changing it with `sudo passwd`, but it still won't work. I also tried changing `PermitRootLogin no` to `PermitRootLogin yes` in my **/etc/ssh/sshd_config**, but it still says "Permission denied, please try again."

**Reply by AyaanZaveri** [reply]
Nevermind got it! I realized I changed the password for the wrong user (user root instead of user coder). 

![Screen Shot 2021-08-19 at 3 44 53 PM](https://user-images.githubusercontent.com/63752541/130134060-e8dd8b20-ca72-484b-9ec5-47b4053d807a.png)

<img width="1440" alt="Screen Shot 2021-08-19 at 3 52 48 PM" src="https://user-images.githubusercontent.com/63752541/130135189-458fa215-951e-4a4c-a185-282da6d32cb3.png">

**Reply by lg-nineoneninethreenine** [reply]
Well, this is VERY VERY interesting. Can you explain me and @jsjoeio how you did it? Wanna try it that's why.

**Reply by AyaanZaveri** [reply]
# Steps

```
sudo apt update

sudo apt install wget

sudo apt install unzip

sudo apt install openssh-server

sudo service ssh start

sudo passwd coder
```
---
```
wget "https://bin.equinox.io/c/4VmDzA7iaHb/ngrok-stable-linux-amd64.zip"

unzip "ngrok-stable-linux-amd64.zip"

./ngrok authtoken YOUR_AUTHTOKEN

./ngrok tcp 22
```
---
Copy the forwarded link (example: tcp://0.tcp.ngrok.io:12345)

Then `ssh coder@0.tcp.ngrok.io -p 12345` and put your password.

**Reply by lg-nineoneninethreenine** [reply]
Oh wow! I will test it later

**Reply by AyaanZaveri** [reply]
Thanks!

**Reply by lg-nineoneninethreenine** [reply]
Okay, made the deployment and I am currently testing this! Thanks @AyaanZaveri and maybe @jsjoeio why not add this to the guides for the people wanting SSH for deployed version of code-server? Can be helpful, right? Or shall I do it?

**Reply by lg-nineoneninethreenine** [reply]
:boom: IT WORKS!!!!!!!!! At my test code-server, I installed ngrok, and forwarded the port. Then, I opened Git Bash and SUCCESSFULLY connected!!!!!!

**Reply by AyaanZaveri** [reply]
Congrats!!!

**Reply by AyaanZaveri** [reply]
I just found a much better way with Cloudflares's Argo Tunnels.


**Install**
```
sudo wget https://github.com/cloudflare/cloudflared/releases/latest/download/cloudflared-linux-amd64.deb

sudo dpkg -i ./cloudflared-linux-amd64.deb

cloudflared tunnel --url ssh://localhost:22
```

**Config**

Download cloudflared from "https://developers.cloudflare.com/cloudflare-one/connections/connect-apps/install-and-setup/installation" for your local computer

Then, go to ~/.ssh/config and add the following

```
Host *.trycloudflare.com
	HostName %h
	User root
	Port 22
	ProxyCommand "cloudflared location" access ssh --hostname %h
```


**SSH**
ssh coder@https://your-link.trycloudflare.com

or 

ssh coder@your-link.trycloudflare.com

**Reply by lg-nineoneninethreenine** [reply]
I will test that too. Thanks for all that help and let's see if I or someone gets permit to add it to the documentation.

**Reply by AyaanZaveri** [reply]
@jsjoeio Can you add this to the documentation?

**Reply by jsjoeio** [reply]
Wow, @AyaanZaveri - fantastic work!!! This is amazing. Yes, I'll open an issue to document this.

**Reply by AyaanZaveri** [reply]
@jsjoeio Thank you so much!!!

**Reply by jsjoeio** [reply]
No, thank YOU for not giving up, figuring out a solution and sharing it with us! 🙌

https://github.com/cdr/code-server/issues/4017

**Reply by AyaanZaveri** [reply]
If you want to create a quick ssh tunnel you can add a task in your tasks.json `/home/coder/.local/share/code-server/User/tasks.json` 
:

 ```
 {
      "label": "ssh",
      "type": "shell",
      "command": "cloudflared tunnel --url ssh://localhost:22",
      "presentation": {
        "reveal": "always"
      },
      "problemMatcher": [],
      "options": {
        "statusbar": {
          "label": "$(broadcast) cloudflare: ssh"
        }
      }
```
<img width="315" alt="Screen Shot 2021-08-23 at 3 24 13 PM" src="https://user-images.githubusercontent.com/63752541/130507164-b70095fd-fc6e-4efe-8247-47cf3b0b0022.png">

**Reply by jsjoeio** [reply]
cc @bpmct you might find this all interesting

**Reply by lg-nineoneninethreenine** [reply]
Does Coder in the cloud have OpenSSH-Server built-in and up and running?

**Reply by jsjoeio** [reply]
I'm not sure! That would be a question for @bpmct

**Reply by jsjoeio** [reply]
> when using sysbox, we run systemd which starts opensshd

Okay this is the response I got when talking to the engineers who work on Coder (the product).

**Reply by lg-nineoneninethreenine** [reply]
Ooh, so now, I learnt how to get Coder up and running, and now, I can use SSH to access it. This is really cool.
